src/app/commissioner_app.cpp

```diff
@@ -152,7 +152,7 @@ Error CommissionerApp::Start(const std::string &aBorderAgentAddr, uint16_t aBord
 
 void CommissionerApp::Stop()
 {
-    mCommissioner->Resign();
+    IgnoreError(mCommissioner->Resign());
```

Comment:
We have a `IgnoreReturnValue()` macro.  Can we consider using/combining with that?

Reply:
Just noticed we don't have this `IgnoreReturnValue()` in this repo. It is in OT main.

Reply:
even in repo `openthread/openthread`, I think `IgnoreError` makes it clearer that we are ignoring  a error. Of cause, we can define `IgnoreError` to `IgnoreReturnValue`.

Reply:
In some cases, I believe `IgnoreReturnValue` is more descriptive. For example, in cases where the error is passed via a parameter. Thoughts?

Reply:
@jwhui, Yes `IgnoreError` is not to be used for any return values. It is designed to be used for `Error` / `otError` exclusively.

Reply:
I think `IgnoreReturnValue` is more descriptive of its intended purpose - to ignore the return value. It's not clear to me why we need to define a separate macro.

Reply:
What I am thinking is:

Typically, there are two kinds of return value: error code and normal data. When readers see `IgnoreError(...)`, they know we are ignoring any error of a function call and we allow errors at there. While `IgnoreReturnValue(...)` doesn't give the information of what is being ignored (error or normal data), so the reader doesn't know if we allow a failure at there untill he checks the function signature by himself.

Current usages of `IgnoreReturnValue` in openthread are all applied on functions return `otError`, **I think, because only error is something deserves an explicit ignoring**. For example, `Message::GetLength()` always success and return the length of the message, there is no meaning for the caller to explicitly ignore the return value, because if the return value is not needed, why should the function be called? 

There are third-party libraries that return error code and normal data in a single signed integer, but I don't think it is a good practice in openthread code. If we need to check the return value of a function against some constants or boundaries to verify if the function call succeed, then we should make the function return `otError` and pass the "true" return value through a by-reference parameter. While we can still use `IgnoreReturnValue` on those thrid-party APIs.

Thoughts?

Reply:
I would also favor using one (more generic) macro `IgnoreReturnValue`.

> Current usages of IgnoreReturnValue in openthread are all applied on functions return otError, I think, because only error is something deserves an explicit ignoring.

I think we already have cases where we want to ignore return values that are not necessarily an error.  Example of this is `Message::Read()` which returns number of bytes read from a message. If the developer knows already that the number of bytes that are being read is smaller than available length in the message, it is perfectly reasonable to ignore the return value.  Example of this in [mesh_forwarder_ftd.cpp](https://github.com/openthread/openthread/blob/7456be9e80b230b734ad454d13243b2978e3364e/src/core/thread/mesh_forwarder_ftd.cpp#L238-L250).

Reply:
@abtink Thanks for your example. My overall idea is making `Error` and `otError` special, but not generic (**failure and return value are totally two different things I think**). In high-level languages, they use `exception` for failures and `exception` is not ignorable. I here want to mimic the semantic of `exception` with `otError` and make it also not ignorable (since we have already do so with macros `VerifyOrExit` and `SuccessOrExit`). To force using `IgnoreError` on only `otError`, we can even have it as a function:
```c++
static inline void IgnoreError(otError aError)
{
    (void)aError;
}
```

Google doesn't allow exception in C++ code and we have [abseil::Status](https://github.com/abseil/abseil-cpp/blob/master/absl/status/status.h) (abseil is the opensource version of the C++ fundamental library used inside Google) to represent failures (or success).

@jwhui @abtink If you agree on making `otError` special than general return value, I think `IgnoreError` is better than `IgnoreReturnValue`. Otherwise, I am still Ok with using `IgnoreReturnValue`.

Reply:
`otError` and `Error` are used extensively in `openthread` and `ot-commissioner` code, making it worthy to be treated differently from other return values. 
I think using `IgnoreError` has following advantages over using `IgnoreReturnValue`:

- `IgnoreError` tells the code reader explicitly that what we are ignoring here is an `Error`, otherwise the reader could wonder what's the type of the return value that is ignored here? He/she will have to check the prototype of the called function to make it clear.
- In the case when `otError` is passed as a function argument, we can also use `IgnoreError` instead of `OT_UNUSED_VARIABLE` to explicit convey our intention to ignore the `Error`. 
- `IgnoreError` is shorter than `IgnoreReturnValue`.

Overall, I think the benefits of using `IgnoreError` outweigh the cost.

Reply:
I fully agree not allowing returned `otError` from methods/function to be ignored would be a useful change. It ensures that developer to consider each situation (either handle the error or decide it is safe to ignore it). 

Having an `IgnoreError()` is fine with me, but we still need to have `IgnoreReturnValue` for other cases (where return value is not `otError` and would be useful for developer to indicate she/he is purposely ignoring it).

One thing to be careful is that  `otError` use in OT does not fully equate to the notion of exceptions. Exceptions typically indicate  an unexpected/rare failure (in bad or unrecoverable way) and they are not expected generally to happen as part of the normal flow of a program. The `otError` enumeration on other hand is used to convey "status" in many cases. Getting an `otError` (which is not an `OT_ERROR_NONE`) in many situations is common or even part of the normal flow of the code. Good examples of this are methods/APIs that iterate over some list and they return `OT_ERROR_NOT_FOUND` when iterator reaches end of the list. Getting this error  is expected as part of the normal flow when going through the list.

Reply:
@abtink Thanks for your explanation of `otError` usage! Agree that we cannot take `otError` the same thing as an Exception since normal flow may depend on the error code. As `otError` is the only representation of failure (at least it should be?) and in most cases in represents for only failure (of cause it can be used for normal flow in specific cases), I believe it deserves a separate macro/function to handler it, as either an return value or function parameter.